resources/providers/awslib/config_provider.go

```diff
@@ -18,23 +18,29 @@
 package awslib
 
 import (
-	"github.com/aws/aws-sdk-go-v2/aws/external"
+	"context"
+	awssdk "github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/elastic/beats/v7/x-pack/libbeat/common/aws"
 )
 
 type ConfigGetter interface {
-	GetConfig() Config
+	InitializeAWSConfig(ctx context.Context, cfg aws.ConfigAWS) (awssdk.Config, error)
```

Comment:
A few points about this interface:
- it only provides the AWS config, so it's not generic as the name suggests
- ideally, the interface and the implementation should be kept in separate files
- the implementation name is not consistent with the interface (getter vs provider)

I would expect each FW to implement the interface on its own level. This will also will allow us to avoid awkward naming situations like this. So instead of `ConfigProvider` that implements the `ConfigGetter` interface for the AWS use case, I would suggest a more general interface `ConfigProvider`, that will have `InitializeConfig` function instead of specific InitializeAWSConfig, and `AwsConfigProvider` will implement it.

Reply:
> If I would think that this interface will be common to a few structs, my answer would have been deferred.

This is exactly the case I'm referring to, a generic provider interface for different providers, such as the EKS. I wouldn't expect to have all of the implementations in the same file as the interface.

> > the implementation name is not consistent with the interface (getter vs provider)
> 
> Totally agree - 💯 . Do you have a suggestion?

I would stick to _provider_

> Sounds like a good idea.
But I wonder, what is mutual to every config provider? 
The awsConfig provider will return different properties than the azureConfig provider and the GCPConfig provider and so on.
So if the **output** of the interface and the **input** of the interface will be the diffenert, we will use `interface{}` as the type of both input and output.
And then I wonder if there is any point in using the generalization.

They provide `Config` which can be extended. If they don't have anything in common, then I think there's no need for an interface in the first place, right?, just implement AwsConfigProvider.

Reply:
I think the usage of interfaces in these kinds of cases is very important since we are using third-party code (AWS libraries).

The interface creates segregation between the concrete AWS lib implementation (interface segregation principle) and it makes our code more testable.
The usage of an interface here, let me mock the provider so I'll be able to test the config provider and the ECR/ELB/EKS - factories.
Without using the interface it will enforce the test use the real AWS ec2 metadata provider.

Other than that, using interfaces makes our code less coupled to any concrete implementation.

Reply:
the reasons behind using an interface here are clear to me, my point was about the use of a generic one, i.e. `configProvider` versus a specific one, i.e. `AwsConfigProvider`, sorry for not being clear.